bayes_opt/acquisition.py

```diff
@@ -0,0 +1,854 @@
+"""Acquisition functions for Bayesian Optimization."""
+import warnings
+import numpy as np
+from numpy.random import RandomState
+from scipy.optimize import minimize
+from scipy.stats import norm
+from scipy.special import softmax
+from .target_space import TargetSpace
+from .constraint import ConstraintModel
+from sklearn.gaussian_process import GaussianProcessRegressor
+from typing import Callable, List, Union, Tuple
+from copy import deepcopy
+from numbers import Number
+
+class ConstraintNotSupportedError(Exception):
+    """Raised when constrained optimization is not supported."""
+
+    pass
+
+
+class NoValidPointRegisteredError(Exception):
+    """Raised when an acquisition function depends on previous points but none are registered."""
+
+    pass
+
+
+class TargetSpaceEmptyError(Exception):
```

Comment:
I'm not entirely sure whether it is appropriate to define such custom errors, or whether it would be better to use `ValueError`s and `TypeError`s here.

Reply:
I'm not sure either actually. I've been wondering this recently myself.

On the plus side, custom errors tend to be clearer. However, for an external program/user to catch these exceptions can be annoying. 

After a bit of reading, I think that if we expect that an external program may want to handle the errors then using standard errors may be better - but if 
1. the error would be expected to be handled internally
2. OR not handled at all (that is: crash is the desired result), 
3. OR there is no standard error that can reasonably be applied
THEN a custom error may be better?

- https://softwareengineering.stackexchange.com/questions/381893/is-it-a-good-practice-to-use-self-defined-exception
- https://stackoverflow.com/questions/45735280/how-to-catch-custom-exception-in-python

Reply:
Thanks for the links! Some good answers in there.
My personal conclusion is this: I think the goal here is to crash the program with an informative error, so that the user knows what to fix. Every single one of these Errors have a clear fix, so I think it's reasonable to have custom classes here and not to conflate them with something potentially more abstract. I would lean to keeping them, though I don't have strong feelings.